app/controllers/session_controller.rb

```diff
@@ -612,4 +642,49 @@ def render_sso_error(status:, text:)
   def sso_url(sso)
     sso.to_url
   end
+
+  # the invite_key will be present if set in InvitesController
+  # when the user visits an /invites/xxxx link; however we do
+  # not want to complete the SSO process of creating a user
+  # and redeeming the invite if the invite is not redeemable or
+  # for the wrong user
+  def validate_invitiation!(sso)
+    invite_key = secure_session["invite-key"]
+    return if invite_key.blank?
+
+    @invite = Invite.find_by(invite_key: invite_key)
+
+    if @invite.blank?
+      raise Invite::ValidationFailed.new(I18n.t("invite.not_found", base_url: Discourse.base_url))
+    end
+
+    if @invite.redeemable?
+      if !@invite.is_invite_link? && sso.email != @invite.email
+        raise Invite::ValidationFailed.new(I18n.t("invite.not_matching_email"))
+      end
+    else
+      if @invite.expired?
+        raise Invite::ValidationFailed.new(I18n.t('invite.expired', base_url: Discourse.base_url))
+      elsif @invite.redeemed?
+        raise Invite::ValidationFailed.new(I18n.t('invite.not_found_template', site_name: SiteSetting.title, base_url: Discourse.base_url))
+      end
+    end
+  end
+
+  def redeem_invitation(sso)
+    InviteRedeemer.new(
+      invite: @invite,
+      username: sso.username,
+      name: sso.name,
+      ip_address: request.remote_ip,
+      session: session,
+      email: sso.email
+    ).redeem
+    secure_session["invite-key"] = nil
+
+  # note - more specific errors are handled in the sso_login method
+  rescue ActiveRecord::RecordInvalid, ActiveRecord::RecordNotSaved => e
+    Rails.logger.warn("SSO invite redemption failed: #{e}")
+    raise Invite::RedemptionFailed
```

Comment:
When invite redemption fails, should we reset the `invite-key` in the session?

Reply:
I think we should only clear it if the redemption passes. Otherwise if you refresh the page it goes through a completely different SSO flow

Reply:
👍